### ISSUE TITLE
zprint doesn't exit cleanly when running via clj

### ISSUE THREAD

**Issue by eerohele** [kkinnear/zprint#107]
This is a bit of a weird one.

Given:

```clojure
;; deps.edn
{:paths ["src"]

 :deps  {org.clojure/clojure {:mvn/version "1.10.0"}
         zprint              {:mvn/version "0.4.15"}}}
```

And:

```clojure
;; src/zprint_hang/core.clj
(ns zprint-hang.core
  (:require [zprint.core :as zprint]))

(defn -main
  [& _]
  (zprint/zprint {:a [:b :c]}))
```

Then:

```bash
$ brew info clojure
clojure: stable 1.10.0.442
$ clj -m zprint-hang.core
{:a [:b :c]}
```

Running `clj` prints the output but doesn't exit unless you press <kbd>Ctrl</kbd> + <kbd>C</kbd>.

However, if you remove `:c` from the vector, the command exits cleanly.

I'm unsure whether this is actually related to `clj`, but using it is the easiest way to reproduce the error, I think.

I'm using macOS.

**Comment by kkinnear** [comment]
Great question! 

In order to maximize performance, zprint uses multiple threads to do the formatting.  Since zprint uses `future` for the multithreading, there is a thread pool used which will cause the program to hang for around a minute on exit if you don't call `(shutdown-agents)`.  In fact, your example will exit in about a minute.  But who wants to wait?

You can do one of two things when calling zprint in your code to prevent this problem:

1. You can tell zprint to not use multiple threads to format at some cost to performance.  Unless you are formatting something very large (or very deeply nested), you might not notice any performance loss.  In that case, your example would be:
```clojure
(ns zprint-hang.core
  (:require [zprint.core :as zprint]))

(defn -main
  [& _]
  (zprint/zprint {:a [:b :c]} {:parallel? false}))
```
An alternative way to do this same thing, but not have to specify an options map on each call:
```clojure
(ns zprint-hang.core
  (:require [zprint.core :as zprint]))

(defn -main
  [& _]
  (zprint/set-options! {:parallel? false})
  (zprint/zprint {:a [:b :c]}))
```
2. You can let zprint run multiple threads, but tell Clojure to shutdown the thread pool before exiting:
```clojure
(ns zprint-hang.core
  (:require [zprint.core :as zprint]))

(defn -main
  [& _]
  (zprint/zprint {:a [:b :c]})
  (shutdown-agents))
```
Either of these should allow your program to exit cleanly.

**Comment by eerohele** [comment]
Thank you for the comprehensive answer!

That makes perfect sense. I tried to dig into the code a bit to figure it out, but I guess I didn't get as far as `:parallel?`, which might've tipped me off.

In my case, option #1 fits the bill perfectly. Thanks again!

**Comment by kkinnear** [comment]
Actually, I went off and read the README after I hit the "COMMENT" button too, and it reminded me that zprint isn't supposed to be using multiple threads when called as a library, for precisely the reason that you ran into.  Something about the `cli` invocation is triggering the "use multiple threads at a REPL" code, so I will look into that.  But until I figure that out, you have a workaround.  Thanks for submitting this issue so that I know to figure out what's going on with `cli`.

**Comment by kkinnear** [comment]
Thanks again for submitting this.  This issue has nothing to do with `clj` invocations.  My code to detect when running in a REPL was just wrong.  After searching around, I have found some better (though still quite empiric) code to detect when running in a REPL and allow parallel operation.  So, after version `0.4.16` you shouldn't have to set `:parallel?` to `false`.  But it will never hurt to do so.

**Comment by kkinnear** [comment]
This is actually fixed in version 0.4.16, with better (i.e. correct) detection of when zprint is running in a REPL.
